File: PHP/DataStructures/BinarySearchTree.php
<?php

class BinarySearchTree {
    private $root;

    public function isEmpty() {
        return $this->root == null;
    }
    // Insert a new element into the BST.
    public function add($value) {
        if($this->root == null) {
            $this->root = new Node($value, null, null);
            return;
        }
        $this->addAux($value, $this->root);
    }

    // Private helper method for the add function.
    private function addAux($value, $treeNode) {
        // If the value is greater, go to the right subtree.
        if($value > $treeNode->value) {
            if($treeNode->right != null) $this->addAux($value, $treeNode->right);
            else $treeNode->right = new Node($value, null, null);
            
        }
        // Otherwise attempt the left subtree.
        elseif($value < $treeNode->value) {
            if($treeNode->left != null) $this->addAux($value, $treeNode->left);
            else $treeNode->left = new Node($value, null, null);

        }
        // If equal, do nothing.
        
    }

    // If the BST contains a specific value.
    public function contains($value){
        if($this->root == null) return false;
        else return $this->containsAux($value, $this->root);
    }

    // Helper method for finding if a BST contains a value.
    private function containsAux($value, $treeNode) {
        if($treeNode == null) return false;

        if($value > $treeNode->value) return $this->containsAux($value, $treeNode->right);
        elseif ($value < $treeNode->value) return $this->containsAux($value, $treeNode->left);
        else return true;
    }

    // Like contains, finds a value in the BST and returns it.
    public function find($value) {
        if($this->root == null) return null;
        else return $this->findAux($value, $this->root);
    }

    // Helper method for the find function.
    private function findAux($value, $treeNode) {
        if($treeNode == null) return null;

        if($value > $treeNode->value) return $this->findAux($value, $treeNode->right);
        elseif($value < $treeNode->value) return $this->findAux($value, $treeNode->left);
        else return $treeNode;
    }

    // Remove a value from a BST.
    public function remove($value) {
        if($this->root == null) return;
        else $this->root = $this->removeAux($value, $this->root);
    }

    // Helper method for the remove function that does most of the heavy lifting.
    private function removeAux($value, $treeNode) {
        if($treeNode == null) return null;

        // Traversal start.
        if($value > $treeNode->value) {
            $treeNode->right = $this->removeAux($value, $treeNode->right);
        }
        elseif($value < $treeNode->value) {
            $treeNode->left = $this->removeAux($value, $treeNode->left);
        }
        // Value found, Now remove;
        else {
            // If both subtrees are not null, 
            // find the smallest element int the right subtree.
            if($treeNode->left != null && $treeNode->right != null) {
                $tmp = $treeNode;
                $leftMost = $this->getLeftMost($tmp->right);

                $treeNode->value = $leftMost->value;
                $treeNode->right = $this->removeAux($leftMost->value, $treeNode->right);
            }
            // If the left subtree isn't null (and the right is), set the left.
            elseif ($treeNode->left != null) {
                $treeNode = $treeNode->left;
            }
            // Else if the right subtree isn't null and the left is.
            elseif ($treeNode->right != null) {
                $treeNode = $treeNode->right;
            }
            // Else no subtrees exist and this node can be set to null.
            else {
                $treeNode = null;
            }
        }
        return $treeNode;
    }

    // Helper method to get the left-most node of a subtree.
    private function getLeftMost($treeNode) {
        if($treeNode == null) return null;
        elseif ($treeNode->left == null) return $treeNode;
        else return $this->getLeftMost($treeNode->left);
    }

    // Clear out the BST.
    public function clear() {
        $this->root = null;
    }

    // Convert the BST to a string.
    public function __toString() {
        $res = "{ " . $this->toStringAux($this->root) . "}";
        return $res;
    }

    // Helper method for the toString function.
    private function toStringAux($treeNode) {
        $res = "";
        
        if(!$treeNode) {
            return "";
        }
        
        if($treeNode->left != null) {
            $res = $res . $this->toStringAux($treeNode->left);
        }
        $res .= (string) $treeNode->value . " ";
        if($treeNode->right != null) {
            $res .= $this->toStringAux($treeNode->right);
        }

        return $res;
    }
    
}

// Node class for the BST that contains a value and a left + right subtree.
class Node {
    
    public $value;
    public $right;
    public $left;

    public function __construct($val, $rightNode, $leftNode){   
        $this->value = $val;
        $this->right = $rightNode;
        $this->left = $leftNode;
    }
}

/* Example usage of the BST */
$bst = new BinarySearchTree();
$bst->add(5);
$bst->add(3);
$bst->add(7);
$bst->add(1);
$bst->add(4);
echo $bst->toString() . "<br>";
$bst->remove(3) . "<br>";
echo $bst->toString() . "<br>";

File: CPlusPlus/DataStructures/BinarySearchTree.cpp
#include <stdio.h>

class Node {
    public:
        Node(int value)
            : value(value), left(nullptr), right(nullptr)
        {}
        int value;
        Node* right;
        Node* left;
};

class BinarySearchTree {
    public:
        BinarySearchTree()
            : root(nullptr)
        {}
        ~BinarySearchTree();
        bool isEmpty();
        void add(int value);
        bool contains(int value);
        Node* find(int value);
        void remove(int value);
        void clear();
    
    private:
        void addAux(int value, Node* treeNode);
        bool containsAux(int value, Node* treeNode);
        Node* findAux(int value, Node* treeNode);
        Node* removeAux(int value, Node* treeNode);
        Node* getLeftMost(Node* treeNode);
        void freeHelper(Node* treeNode);
        Node* root;
};

BinarySearchTree::~BinarySearchTree() {
    this->freeHelper(this->root);
    this->root = nullptr;
}

void BinarySearchTree::freeHelper(Node* treeNode) {
    if(treeNode == nullptr) {
        return;
    }
    if(treeNode->left != nullptr) {
        this->freeHelper(treeNode->left);
    }
    if(treeNode->right != nullptr) {
        this->freeHelper(treeNode->right);
    }
     
    delete treeNode;
}

bool BinarySearchTree::isEmpty() {
    return this->root == nullptr;
}

void BinarySearchTree::add(int value) {
    if(this->root == nullptr) {
        root = new Node(value);
        return;
    }

    this->addAux(value, root);
}


void BinarySearchTree::addAux(int value, Node* treeNode) {
    if(value > treeNode->value) {
        if(treeNode->right != nullptr) this->addAux(value, treeNode->right);
        else treeNode->right = new Node(value);
    }
    else if(value < treeNode->value) {
        if(treeNode->left != nullptr) this->addAux(value, treeNode->left);
        else treeNode->left = new Node(value);
    }
    else {
        return;
    }
}

bool BinarySearchTree::contains(int value) {
    return this->containsAux(value, this->root);
}

bool BinarySearchTree::containsAux(int value, Node* treeNode) {
    if(treeNode == nullptr) return false;

    if(value > treeNode->value) return this->containsAux(value, treeNode->right);
    else if(value < treeNode->value) return this->containsAux(value, treeNode->left);
    else return true;
}

Node* BinarySearchTree::find(int value) {
    return this->findAux(value, this->root);
}

Node* BinarySearchTree::findAux(int value, Node* treeNode) {
    if(treeNode == nullptr) return nullptr;

    if(value > treeNode->value) return this->findAux(value, treeNode->right);
    else if(value < treeNode->value) return this->findAux(value, treeNode->left);
    else return treeNode;
}

void BinarySearchTree::remove(int value) {
    root = this->removeAux(value, root);
}

Node* BinarySearchTree::removeAux(int value, Node* treeNode) {
    if(treeNode == nullptr) return nullptr;

    if(value > treeNode->value) {
        treeNode->right = this->removeAux(value, treeNode->right);
    }
    else if(value < treeNode->value) {
        treeNode->left = this->removeAux(value, treeNode->left);
    }
    else {
        if(treeNode->left != nullptr && treeNode->right != nullptr) {
            Node* tmp = treeNode;
            Node* leftMost = this->getLeftMost(tmp->right);
            treeNode->value = leftMost->value;

            treeNode->right = this->removeAux(leftMost->value, treeNode->right);
        }
        else if(treeNode->left != nullptr) {
            Node* deadPtr = treeNode;
            treeNode = treeNode->left;
            delete deadPtr;
        }
        else if(treeNode->right != nullptr) {
            Node* deadPtr = treeNode;
            treeNode = treeNode->right;
            delete deadPtr;
        }
        else {
            delete treeNode;
            treeNode = nullptr;
        }
    }
    return treeNode;
}

Node* BinarySearchTree::getLeftMost(Node* treeNode) {
    if(treeNode == nullptr) return nullptr;
    else if(treeNode->left == nullptr) return treeNode;
    else return this->getLeftMost(treeNode->left);
}
